src/main/java/conseq4j/Conseq.java

```diff
@@ -1,192 +1,289 @@
+                new Semaphore(builder.globalConcurrency, FIFO_ON_CONCURRENCY_CONTENTION),
+                builder.executorTaskQueueCapacity), genericObjectPoolConfig);
+    }
 
+    @Override
+    public void execute(Object sequenceKey, Runnable runnable) {
+        sequentialExecutors.compute(sequenceKey, (presentSequenceKey, presentExecutor) -> {
+            GlobalConcurrencyBoundedRunningTasksCountingExecutorService computed = computeExecutor(presentSequenceKey,
+                    presentExecutor);
+            computed.execute(runnable);
+            return computed;
+        });
     }
 
-    /**
-     * @return Max count of concurrent executors
-     */
-    public int getMaxConcurrentExecutors() {
-        return this.consistentHasher.getTotalBuckets();
+    private GlobalConcurrencyBoundedRunningTasksCountingExecutorService computeExecutor(Object presentSequenceKey,
+            GlobalConcurrencyBoundedRunningTasksCountingExecutorService presentExecutor) {
+        if (presentExecutor != null) {
+            return presentExecutor;
+        }
+        final GlobalConcurrencyBoundedRunningTasksCountingExecutorService computed;
+        try {
+            computed = this.executorPool.borrowObject();
+        } catch (Exception ex) {
+            throw new IllegalStateException("Failed to borrow executor from pool " + this.executorPool, ex);
+        }
+        computed.addListener(new SweepingExecutorServiceListener(presentSequenceKey, sequentialExecutors,
+                executorPool));
+        return computed;
     }
 
-    int getSingleExecutorTaskQueueSize() {
-        return this.sequentialExecutorServiceCacheLoader.getExecutorQueueSize();
+    @Override
+    public <T> Future<T> submit(Object sequenceKey, Callable<T> task) {
+        FutureHolder<T> futureHolder = new FutureHolder<>();
+        sequentialExecutors.compute(sequenceKey, (presentSequenceKey, presentExecutor) -> {
+            GlobalConcurrencyBoundedRunningTasksCountingExecutorService computed = computeExecutor(presentSequenceKey,
+                    presentExecutor);
+            futureHolder.set(computed.submit(task));
+            return computed;
+        });
+        return futureHolder.get();
     }
 
     @Override
-    public ExecutorService getSequentialExecutor(CharSequence sequenceKey) {
-        return this.executorCache.get(this.consistentHasher.hashToBucket(sequenceKey));
+    public <T> Future<T> submit(Object sequenceKey, Runnable task, T result) {
+        FutureHolder<T> futureHolder = new FutureHolder<>();
+        sequentialExecutors.compute(sequenceKey, (presentSequenceKey, presentExecutor) -> {
+            GlobalConcurrencyBoundedRunningTasksCountingExecutorService computed = computeExecutor(presentSequenceKey,
+                    presentExecutor);
+            futureHolder.set(computed.submit(task, result));
+            return computed;
+        });
+        return futureHolder.get();
```

Comment:
You can't return futureHolder.get() As it is async call, that value would still be null. But even if you return future, you wouldn't know when the computation is finished and future is populated.

Reply:
good point. thx!. actually, i need to make the FutureHolder to the same thing as the Future to the API client: including things like isDone() and blocking get(). no?

Reply:
correct i think those two methods are mandatory here. Or a method inside the future that can append next async job on that future. Like [Vertx ](https://vertx.io/) futures are working. But i guess then there are new, even harder problems to solve :)